File: belgian-monitor-parser.js
'use strict';

const jsdom = require('jsdom')
const moment = require('moment')
const url = require('url')

const buildUrl = (enterpriseNumber) => `http://www.ejustice.just.fgov.be/cgi_tsv/tsv_l_1.pl?lang=fr&sql=btw+contains+%27${enterpriseNumber}%27&fromtab=TSV&rech=1&pdda=&pddm=&pddj=&pdfa=&pdfm=&pdfj=&naam=&postkode=&localite=&numpu=&hrc=&akte=&btw=${enterpriseNumber}&jvorm=&land=&set2=&set3=`

const searchDocuments = (enterpriseNumber) => new Promise( (resolve, reject) => {
  const searchUrl = buildUrl(enterpriseNumber)
  console.log(`URL: ${searchUrl}`)
  jsdom.env({
    url: searchUrl,
    done: (err, window) => {
      if(err) {
        reject(err)
      }
      const docs = Array.prototype.slice.call(window.document.querySelectorAll('center table tr td:nth-child(2)'))
      console.log(`Number of documents found: ${docs.length}`)
      const documentsList = docs.map( (doc) => {
          const parts = doc.innerHTML.split('<br>').slice(3)
          const title = parts[0].trim()
          const linkNode = doc.querySelector('a')
          const link = linkNode ? url.resolve(searchUrl, linkNode.getAttribute('href')) : null
          const publishedAt = moment(parts[1].slice(0, 10)).toDate()
          return { title, publishedAt, link }
      })
      resolve(documentsList)
    }
  })
})

module.exports = { searchDocuments }

File: belgian_monitor_parser.rb
require 'nokogiri'
require 'open-uri'
require 'uri'

class BelgianMonitorParser

  def self.normalize_enterprise_number enterprise_number
    "#{enterprise_number}".gsub(/\D/, "").to_i
  end

  def build_url enterprise_number
    "http://www.ejustice.just.fgov.be/cgi_tsv/tsv_l_1.pl?lang=fr&sql=btw+contains+%27#{enterprise_number}%27&fromtab=TSV&rech=1&pdda=&pddm=&pddj=&pdfa=&pdfm=&pdfj=&naam=&postkode=&localite=&numpu=&hrc=&akte=&btw=#{enterprise_number}&jvorm=&land=&set2=&set3="
  end

  def extract_document_meta doc
    parts = doc.children.reject{ |node| node.name == 'br' }.slice(5..-1)
    title = "#{parts[0]}".strip
    published_at = Date.parse "#{parts[1]}".strip[0..10]
    link_node = doc.at_css('a')
    link = URI.join(@search_url, "#{link_node['href']}") if link_node
    {
      title: title,
      published_at: published_at,
      link: link
    }
  end

  def search_documents enterprise_number
    @search_url = build_url(enterprise_number)
    doc = Nokogiri::HTML(open(@search_url))
    doc.css('center table tr td:nth-child(2)').map(&method(:extract_document_meta))
  end
end
